### ISSUE TITLE
JSON format dataset doesn't work when training LoRA

### ISSUE THREAD

**Issue by Damonlin11** [oobabooga/text-generation-webui#2807]
### Describe the bug

I used the following JSON format for the training data, here is a snippet of the data:
[
    {
        "instruction": "Who are you?",
        "input": "",
        "output": "I am xxxxxx, your personal assistant."
    },
    {
        "instruction": "Comfortable life",
        "input": "",
        "output": "You are in danger of living a life so comfortable and soft that you will die without ever realizing your true potential."
    }
]

But it doesn't work, it raised an error: AttributeError: 'list' object has no attribute 'items'

Is the format I used wrong?

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

I don't know

### Logs

```shell
INFO:root:Loading JSON datasets...
Downloading and preparing dataset json/default to /root/.cache/huggingface/datasets/json/default-25ba5605e4949f70/0.0.0/0f7e3662623656454fcd2b650f34e886a7db4b9104504885bd462096cc7a9f51...
Downloading data files: 100% 1/1 [00:00<00:00, 2240.55it/s]
Extracting data files: 100% 1/1 [00:00<00:00, 71.70it/s]
Dataset json downloaded and prepared to /root/.cache/huggingface/datasets/json/default-25ba5605e4949f70/0.0.0/0f7e3662623656454fcd2b650f34e886a7db4b9104504885bd462096cc7a9f51. Subsequent calls will reuse this data.
100% 1/1 [00:00<00:00, 121.42it/s]
╭───────────────────── Traceback (most recent call last) ──────────────────────╮
│ /content/drive/MyDrive/train-lora-main/do_training.py:196 in <module>        │
│                                                                              │
│   193 │   │   │   shared.args.optimizer,                                     │
│   194 │   │   │   shared.args.hard_cut_string                                │
│   195 │   │   │   ]                                                          │
│ ❱ 196 │   │   do_train(*all_params)                                          │
│   197                                                                        │
│                                                                              │
│ /content/drive/MyDrive/train-lora-main/modules/train_lora.py:204 in do_train │
│                                                                              │
│   201 │   │                                                                  │
│   202 │   │   logging.info("Loading JSON datasets...")                       │
│   203 │   │   data = load_dataset("json", data_files=clean_path('training/da │
│ ❱ 204 │   │   train_data = data['train'].map(generate_and_tokenize_prompt)   │
│   205 │   │                                                                  │
│   206 │   │   if eval_dataset == 'None':                                     │
│   207 │   │   │   eval_data = None                                           │
│                                                                              │
│ /usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py:563 in     │
│ wrapper                                                                      │
│                                                                              │
│    560 │   │   else:                                                         │
│    561 │   │   │   self: "Dataset" = kwargs.pop("self")                      │
│    562 │   │   # apply actual function                                       │
│ ❱  563 │   │   out: Union["Dataset", "DatasetDict"] = func(self, *args, **kw │
│    564 │   │   datasets: List["Dataset"] = list(out.values()) if isinstance( │
│    565 │   │   for dataset in datasets:                                      │
│    566 │   │   │   # Remove task templates if a column mapping of the templa │
│                                                                              │
│ /usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py:528 in     │
│ wrapper                                                                      │
│                                                                              │
│    525 │   │   │   "output_all_columns": self._output_all_columns,           │
│    526 │   │   }                                                             │
│    527 │   │   # apply actual function                                       │
│ ❱  528 │   │   out: Union["Dataset", "DatasetDict"] = func(self, *args, **kw │
│    529 │   │   datasets: List["Dataset"] = list(out.values()) if isinstance( │
│    530 │   │   # re-apply format to the output                               │
│    531 │   │   for dataset in datasets:                                      │
│                                                                              │
│ /usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py:2953 in    │
│ map                                                                          │
│                                                                              │
│   2950 │   │   │   │   │   leave=False,                                      │
│   2951 │   │   │   │   │   desc=desc or "Map",                               │
│   2952 │   │   │   │   ) as pbar:                                            │
│ ❱ 2953 │   │   │   │   │   for rank, done, content in Dataset._map_single(** │
│   2954 │   │   │   │   │   │   if done:                                      │
│   2955 │   │   │   │   │   │   │   shards_done += 1                          │
│   2956 │   │   │   │   │   │   │   logger.debug(f"Finished processing shard  │
│                                                                              │
│ /usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py:3307 in    │
│ _map_single                                                                  │
│                                                                              │
│   3304 │   │   │   │   if not batched:                                       │
│   3305 │   │   │   │   │   _time = time.time()                               │
│   3306 │   │   │   │   │   for i, example in shard_iterable:                 │
│ ❱ 3307 │   │   │   │   │   │   example = apply_function_on_filtered_inputs(e │
│   3308 │   │   │   │   │   │   if update_data:                               │
│   3309 │   │   │   │   │   │   │   if i == 0:                                │
│   3310 │   │   │   │   │   │   │   │   buf_writer, writer, tmp_file = init_b │
│                                                                              │
│ /usr/local/lib/python3.10/dist-packages/datasets/arrow_dataset.py:3210 in    │
│ apply_function_on_filtered_inputs                                            │
│                                                                              │
│   3207 │   │   │   │   additional_args += (effective_indices,)               │
│   3208 │   │   │   if with_rank:                                             │
│   3209 │   │   │   │   additional_args += (rank,)                            │
│ ❱ 3210 │   │   │   processed_inputs = function(*fn_args, *additional_args, * │
│   3211 │   │   │   if isinstance(processed_inputs, LazyDict):                │
│   3212 │   │   │   │   processed_inputs = {                                  │
│   3213 │   │   │   │   │   k: v for k, v in processed_inputs.data.items() if │
│                                                                              │
│ /content/drive/MyDrive/train-lora-main/modules/train_lora.py:199 in          │
│ generate_and_tokenize_prompt                                                 │
│                                                                              │
│   196 │   │   │   raise RuntimeError(f'Data-point "{data_point}" has no keys │
│   197 │   │                                                                  │
│   198 │   │   def generate_and_tokenize_prompt(data_point):                  │
│ ❱ 199 │   │   │   prompt = generate_prompt(data_point)                       │
│   200 │   │   │   return tokenize(prompt)                                    │
│   201 │   │                                                                  │
│   202 │   │   logging.info("Loading JSON datasets...")                       │
│                                                                              │
│ /content/drive/MyDrive/train-lora-main/modules/train_lora.py:190 in          │
│ generate_prompt                                                              │
│                                                                              │
│   187 │   │   │   format_data: dict[str, str] = json.load(formatFile)        │
│   188 │   │                                                                  │
│   189 │   │   def generate_prompt(data_point: dict[str, str]):               │
│ ❱ 190 │   │   │   for options, data in format_data.items():                  │
│   191 │   │   │   │   if set(options.split(',')) == set(x[0] for x in data_p │
│   192 │   │   │   │   │   for key, val in data_point.items():                │
│   193 │   │   │   │   │   │   if val is not None:                            │
╰──────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'list' object has no attribute 'items'
```

### System Info

```shell
GPU T4
```

**Comment by Damonlin11** [comment]
> 

Haha, it is a long story. I used the base code to do the training. I finally got it to work! The dataset format I posted above is correct. The problem is that I didn't tell the module which format the dataset will be matched to. I realized that the parameter --format should tell which format the input dataset will be matched to, and --dataset takes the input dataset.

**Comment by Damonlin11** [comment]
> I want such a fancy traceback! How did you do it? Tell me your secrets.

Haha, it is a long story. I do the training by using some of the base code instead of on webui. I finally got it to work! The dataset format I posted above is correct. The problem is that I didn't tell the module which format the dataset will be matched to. I realized that the parameter --format should tell which format the input dataset will be matched to, and --dataset takes the input dataset.
